File: LeetCode/Java Solutions/Palindrome Linked List/palindrome.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

 class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) return true;
        
        // Find the middle of the list
        ListNode slow = head, fast = head;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        // Reverse the second half
        ListNode secondHalfHead = reverse(slow.next);
        ListNode firstHalfHead = head;
        
        // Compare the first and second half nodes
        boolean result = true;
        while (result && secondHalfHead != null) {
            if (firstHalfHead.val != secondHalfHead.val) result = false;
            firstHalfHead = firstHalfHead.next;
            secondHalfHead = secondHalfHead.next;
        }
        
        // Restore the original list (Optional)
        slow.next = reverse(slow.next);
        
        return result;
    }
    
    private ListNode reverse(ListNode head) {
        ListNode prev = null;
        while (head != null) {
            ListNode nextNode = head.next;
            head.next = prev;
            prev = head;
            head = nextNode;
        }
        return prev;
    }
}

File: LeetCode/Python Solutions/Palindrome Linked List/palindrome.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return True
        
        # Find the middle of the list
        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        
        # Reverse the second half
        second_half_head = self.reverse(slow.next)
        first_half_head = head
        
        # Compare the first and second half nodes
        result = True
        while result and second_half_head:
            if first_half_head.val != second_half_head.val:
                result = False
            first_half_head = first_half_head.next
            second_half_head = second_half_head.next
        
        # Restore the list (optional)
        slow.next = self.reverse(slow.next)
        
        return result
    
    def reverse(self, head: ListNode) -> ListNode:
        prev = None
        while head:
            next_temp = head.next
            head.next = prev
            prev = head
            head = next_temp
        return prev
  